File: Ruby/merge-sort.rb
class Main
  def self.run
    arr = [93, 2, 47, 1, 902]
    puts merge_sort(arr, 0 , arr.length - 1)
  end

  def self.merge_sort(arr, left, right)
    return if left >= right
    middle = (left + right) / 2
    merge_sort(arr, left, middle)
    merge_sort(arr, middle + 1, right)
    merge(arr, middle, left, right)
    arr
  end

  def self.merge(arr, middle, left, right)
    leftArr = arr[(left)..(middle)]
    rightArr = arr[(middle + 1)..(right)]
    i = j = 0
    k = left

    while i < leftArr.length && j < rightArr.length
      if leftArr[i] < rightArr[j] 
        arr[k] = leftArr[i]
        i = i + 1
      else
        arr[k] = rightArr[j]
        j = j + 1  
      end
      k = k + 1
    end

    while i < leftArr.length 
      arr[k] = leftArr[i]
      i = i + 1
      k = k + 1
    end

    while j < rightArr.length 
      arr[k] = rightArr[j]
      j = j + 1
      k = k + 1
    end

  end

end


Main.run

File: C++/merge_sort.cpp
#include<bits/stdc++.h>
using namespace std;
#define N 10

void merge(int A[], int init, int middle, int end){
	int n1 = middle - init + 1;
	int n2 = end - middle;

	int L[n1], R[n2];

	for(int x = 0; x < n1; x++){
		L[x] = A[init + x];
	}
	for(int y = 0; y < n2; y++){
		R[y] = A[middle + y + 1];
	}

	int x = 0, y = 0, k = init;
	while(x < n1 && y < n2){
		if(L[x] <= R[y]){
			A[k] = L[x];
			x++;
		}else{
			A[k] = R[y];
			y++;
		}
		k++;
	}
	while(x < n1){
		A[k] = L[x];
		x++;
		k++;
	}
	while(y < n2){
		A[k] = R[y];
		y++;
		k++;
	}
}

void mergesort(int A[], int init, int end){
	if(init < end){
		int middle = (end + init) / 2;
		mergesort(A, init, middle);
		mergesort(A, middle + 1, end);
		merge(A, init, middle, end);
	}
}

int main(){
	int A[N];
	for(int x = 0; x < N; x++){
		cin >> A[x];
	}
	mergesort(A, 0, N - 1);
	for(int x = 0; x < N; x++){
		cout << A[x] << " ";
	}
}
